### PR TITLE
DRILL-7117: Support creation of equi-depth histogram for selected dat…

### PR DESCRIPTION
**Pull request by amansinha100**
…a types.

- This PR adds support for creating equi-depth histograms on the following data types: INT, BIGINT, FLOAT4, FLOAT8, DATE, TIME, TIMESTAMP and BOOLEAN.   No selectivity calculations have been modified yet (that will be done in a later PR).  

- The histogram is built using the t-digest approximation algorithm and associated data structure.  
Please see details in [DRILL-7117](https://issues.apache.org/jira/browse/DRILL-7117) and the parent JIRA [DRILL-6992](https://issues.apache.org/jira/browse/DRILL-6992) which contains a link to the design document. 

- The same ANALYZE command used for NDV etc will also gather histograms and no new syntax has been added.  For testing, I have done a bunch of manual testing using both skewed and uniform distributions and with different data types.  Please see [DRILL-7117](https://issues.apache.org/jira/browse/DRILL-7117) for results of such testing.  No unit tests have been added yet since the bucket boundaries change slightly by the underlying t-digest.  Making this repeatable and unit-testable needs some thinking and I will do this in a follow-up PR.

### CODE REVIEW

**Comment by amansinha100** [comment]
@gparai could you pls review ?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/TDigestMergedStatistic.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+          TDigest other = TDigest.fromBytes(ByteBuffer.wrap(accessor.get(0)));
+          if (colTdigestHolder != null) {
+            colTdigestHolder.add(other);
+            tdigestHolder.put(colName, colTdigestHolder);
+          } else {
+            tdigestHolder.put(colName, other);
+          }
+        }
+      } catch (Exception ex) {
+        //TODO: Catch IOException/CardinalityMergeException
```
Comment (gparai) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/TDigestMergedStatistic.java]: Missing?

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillStatsTable.java [hunk 2]
```diff
@@ -62,10 +62,15 @@
   public enum STATS_VERSION {V0, V1};
   // The current version
   public static final STATS_VERSION CURRENT_VERSION = STATS_VERSION.V1;
+  // 10 histogram buckets (TODO: can make this configurable later)
+  public static final int NUM_HISTOGRAM_BUCKETS = 10;
```
Comment (gparai) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillStatsTable.java]: Maybe we can look at other implementations to figure out a good default value for histogram buckets.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/TDigestMergedStatistic.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      try {
+        if (colTdigestHolder != null) {
+          int size = colTdigestHolder.smallByteSize();
+          java.nio.ByteBuffer byteBuf = java.nio.ByteBuffer.allocate(size);
+          colTdigestHolder.asSmallBytes(byteBuf);
+          // NOTE: in setting the VV below, we are using the byte[] instead of the ByteBuffer because the
+          // latter was producing incorrect output (after re-reading the data from the VV).  It is
+          // unclear whether the setSafe() api for ByteBuffer has a bug, so to be safe we are using the
+          // byte[] directly which works.
+          vv.getMutator().setSafe(0, byteBuf.array(), 0, byteBuf.array().length);
```
Comment (gparai) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/TDigestMergedStatistic.java]: Sampling will throw off the histogram values just like NDV. Do we plan to address it now or later?

File: exec/jdbc-all/pom.xml [hunk 4]
```diff
@@ -452,6 +452,7 @@
                <exclude>org/apache/drill/shaded/guava/com/google/common/graph/**</exclude>
                <exclude>org/apache/drill/shaded/guava/com/google/common/collect/Tree*</exclude>
                <exclude>org/apache/drill/shaded/guava/com/google/common/collect/Standard*</exclude>
+               <exclude>org/apache/drill/shaded/guava/com/google/common/io/BaseEncoding*</exclude>
```
Comment (gparai) [inline exec/jdbc-all/pom.xml]: Why was this needed?

File: exec/java-exec/src/main/resources/drill-module.conf [hunk 5]
```diff
@@ -685,5 +685,6 @@ drill.exec.options: {
     exec.query.return_result_set_for_ddl: true,
     # ========= rm related options ===========
     exec.rm.queryTags: "",
-    exec.rm.queues.wait_for_preferred_nodes: true
+    exec.rm.queues.wait_for_preferred_nodes: true,
+    exec.statistics.tdigest_compression: 100
```
Comment (gparai) [inline exec/java-exec/src/main/resources/drill-module.conf]: Do we want to expose the algorithm name to the user? We had a similar discussion for HLL where we decided it would be better to have a generic name.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/NumericEquiDepthHistogram.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+@JsonTypeName("numeric-equi-depth")
+public class NumericEquiDepthHistogram implements Histogram {
+
+  // For equi-depth, all buckets will have same (approx) number of rows
+  @JsonProperty("numRowsPerBucket")
+  private long numRowsPerBucket;
+
+  // An array of buckets arranged in increasing order of their start boundaries
+  // Note that the buckets only maintain the start point of the bucket range.
+  // End point is assumed to be the same as the start point of next bucket.
```
Comment (gparai) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/NumericEquiDepthHistogram.java]: I thought end-point was less than the start point of the next bucket?

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/NumericEquiDepthHistogram.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      // get the starting point of the i-th quantile
+      double start = tdigest.quantile(q * i);
+      histogram.buckets[i] = start;
+    }
+    // for the N-th bucket, the end point corresponds to the 1.0 quantile but we don't keep the end
+    // points; only the start point, so this is stored as the start point of the (N+1)th bucket
+    histogram.buckets[i] = tdigest.quantile(1.0);
+
+    // each bucket stores approx equal number of rows
+    histogram.setNumRowsPerBucket(tdigest.size()/numBuckets);
```
Comment (gparai) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/NumericEquiDepthHistogram.java]: `tdigest.size()` would be equal to the sampled rows? In that case we should use the rowcount (which would have done the scaling to account for sampling).

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/StringEquiDepthHistogram.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import org.apache.calcite.rex.RexNode;
+
+/**
+ * A column specific histogram which is meant for string columns
+ */
+@JsonTypeName("string-equi-depth")
+public class StringEquiDepthHistogram implements Histogram {
```
Comment (gparai) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/StringEquiDepthHistogram.java]: Do we want to remove this implementation for now?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/TDigestFunctions.java [hunk 9]
```diff
@@ -0,0 +1,1126 @@
...
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(TDigestFunctions.class);
+
+  private TDigestFunctions(){}
+
+  @FunctionTemplate(name = "tdigest", scope = FunctionTemplate.FunctionScope.POINT_AGGREGATE)
+  public static class BigIntTDigestFunction implements DrillAggFunc {
+    @Param BigIntHolder in;
+    @Workspace ObjectHolder work;
+    @Output NullableVarBinaryHolder out;
+    @Inject DrillBuf buffer;
```
[the remark this answers is not in the document]
Comment (amansinha100) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/TDigestFunctions.java]: The `buffer` is being used in the output() method below.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/TDigestFunctions.java [hunk 10]
```diff
@@ -0,0 +1,1126 @@
...
+      }
+    }
+
+    @Override
+    public void reset() {
+      work.obj = new com.clearspring.analytics.stream.quantile.TDigest(compression.value);
+    }
+  }
+
+  @FunctionTemplate(name = "tdigest", scope = FunctionTemplate.FunctionScope.POINT_AGGREGATE)
```
[the remark this answers is not in the document]
Comment (amansinha100) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/TDigestFunctions.java]: Without this code for Varchar, the code generator for StatisticsAggBatch encounters the following error: 
`0: jdbc:drill:zk=local> analyze table dfs.tmp.nation1 compute statistics;
Error: FUNCTION ERROR: Failure while trying use function. No body found for required method setup.

FunctionClass org.apache.drill.exec.expr.fn.impl.TDigestFunctions$NullableVarCharTDigestFunction
Fragment 0:0
`

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/TDigestFunctions.java [hunk 11]
```diff
@@ -0,0 +1,1126 @@
...
+
+    @Override
+    public void reset() {
+      work.obj = new com.clearspring.analytics.stream.quantile.TDigest(compression.value);
+    }
+  }
+
+  @FunctionTemplate(name = "tdigest_decode", scope = FunctionScope.SIMPLE, nulls = NullHandling.NULL_IF_NULL)
+  public static class TDigestDecode implements DrillSimpleFunc {
+
```
[the remark this answers is not in the document]
Comment (amansinha100) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/TDigestFunctions.java]: Turns out I am not using this function since the 'decode' of the t-digest happens outside the scope of the generated code.  I will remove this function altogether.

File: exec/jdbc-all/pom.xml (hunk 4 above)
Comment (amansinha100) [reply exec/jdbc-all/pom.xml]: The JDBC jar file was exceeding size limits.  We have done this type of exclusions in the past also.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/StringEquiDepthHistogram.java (hunk 8 above)
Comment (amansinha100) [reply exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/StringEquiDepthHistogram.java]: Ok, will remove.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/TDigestMergedStatistic.java (hunk 1 above)
Comment (amansinha100) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/TDigestMergedStatistic.java]: I will add logging for this.

File: exec/java-exec/src/main/resources/drill-module.conf (hunk 5 above)
Comment (amansinha100) [reply exec/java-exec/src/main/resources/drill-module.conf]: The 'compression' is the name that t-digest uses for determining the number of centroids.  I cannot think of something more generic for this since if we change the underlying algorithm it may not use centroids at all.  I am thinking of keeping this as-is.  Let me know if you can think of a generic representation.  We could revisit the HLL names if we want to be consistent.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/NumericEquiDepthHistogram.java (hunk 7 above)
Comment (amansinha100) [reply exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/NumericEquiDepthHistogram.java]: Agree that num rows per bucket should reflect the extrapolated rows, not just the sampled rows.  Even though the t-digest internally stores only the sampled rows and the quantile() function will operate based on that but later when the selectivity calculations are done, they should use the extrapolated num rows per bucket.   I have changed this to use the max(t-digest size,  non-null-rowcount).  The t-digest stores only non-null values,

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/TDigestMergedStatistic.java (hunk 3 above)
Comment (amansinha100) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/TDigestMergedStatistic.java]: A random sample should still be usable for computing the quantiles.  I have changed the numRowsPerBucket based on your other comment elsewhere about sampling.

**Comment by amansinha100** [comment]
@gparai could you take another look ?  I have addressed your review comments.

**Comment by gparai** [comment]
The changes look good. +1
